### PR TITLE
feat: logs in SQL

### PR DESCRIPTION
**Pull request by danielgent**
#### Includes
* a migration script _which has been run_ as needed to integrate this with auth-admin -> https://github.com/sesamyab/auth-admin/pull/58

#### TODO
* double check all logs removed from User model have been put into routes :heavy_check_mark: 
* check that category and message are correct :heavy_check_mark:

### CODE REVIEW

File: src/adapters/in-memory/index.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
 import { DataAdapters } from "../interfaces";
 import { createCertificateAdapter } from "./certificates";
-import { createUserAdapter } from "./user";
+import { createUsersAdapter } from "./user";
```
Comment (danielgent) [inline src/adapters/in-memory/index.ts:3]: fixing plurality here to match others

File: src/adapters/in-memory/user/listUsers.ts [hunk 2]
```diff
@@ -1,10 +1,11 @@
 import { UserResponse, Totals } from "../../../types/auth0";
-import { ListUserParams, ListUsersResponse } from "../../interfaces/User";
+import { ListUsersResponse } from "../../interfaces/Users";
+import { ListParams } from "../../interfaces/ListParams";
```
Comment (danielgent) [inline src/adapters/in-memory/user/listUsers.ts:3]: No difference between `ListUserParams` and `ListParams` so I removed `ListUserParams`

File: src/adapters/interfaces/Logs.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import { Log } from "../../types/";
+import { ListParams } from "./ListParams";
+
+// TODO - remove this Message suffix...
+export interface CreateLogParams {
+  category: string;
+  message: string;
+  tenant_id: string;
+  user_id: string;
+} // this is the same as Omit<Log, "timestamp" | "id"> - any reason to define it?
```
Comment (danielgent) [inline src/adapters/interfaces/Logs.ts:11]: What do you think Markus?

Should we be using `Omit` everywhere? or defining multiple types? I feel like Omit might be better... so we don't have so many duplicate-ish types

File: src/models/User.ts [hunk 4]
```diff
@@ -21,12 +23,7 @@ import { Env, ProfileSchema } from "../types";
 import { sendUserEvent, UserEvent } from "../services/events";
 import { Profile } from "../types";
 import { migratePasswordHook } from "../hooks/migrate-password";
-import { LogMessage } from "../types/LogMessage";
 
-// do these need changing? I'm thinking I've been too aggressive here and these are the DOs which should remain camelcase?
-// BUT if Auth0 does all the user profile (and ONLY the user profile) snake_case everywhere, I think we should copy
-// else we'll end up painting ourselves into a corner
-// TBD
```
Comment (danielgent) [inline src/models/User.ts:29]: my previous rants :smile:

File: test/models/User.spec.ts [hunk 5]
```diff
@@ -14,6 +14,11 @@ function createCaller(storage: DOStorageFixture) {
       USERS_QUEUE: {
         send: async () => {},
       } as unknown as Queue<QueueMessage>,
+      data: {
+        logs: {
+          create: () => {},
+        },
+      } as unknown,
```
Comment (danielgent) [inline test/models/User.spec.ts:21]: minimum change needed so the tests run :smile:

File: src/models/User.ts [hunk 6]
```diff
@@ -119,30 +114,23 @@ async function getEmailValidationCode(storage: DurableObjectStorage) {
...
 
-  logs.push({
+  const { tenant_id, id } = profile;
+
+  await ctx.env.data.logs.create({
     ...message,
-    id: nanoid(),
-    timestamp: new Date().toISOString(),
+    tenant_id,
+    user_id: id,
```
Comment (danielgent) [inline src/models/User.ts:132]: this function here is the main change! Thoughts?

File: src/models/User.ts [hunk 7]
```diff
@@ -431,7 +423,10 @@ export const userRouter = router({
     .mutation(async ({ input, ctx }) => {
       const profile = await updateProfile(ctx, input);
 
-      await writeLog(ctx.state.storage, {
+      console.log("this mostly looks like ctx.env - ", ctx.env);
+      console.log("but this is unpopulated ctx.env.data - ", ctx.env.data);
```
Comment (danielgent) [inline src/models/User.ts:427]: @markusahlstrand I cannot follow the magic here, but the types don't reflect what we actually have here

The rest of `ctx.env` is populated it seems, but `ctx.env.data` is null, so I can't do the writing to log below (with the data adaptor)

Any thoughts?

I suppose we'll need this working even when we drop DOs assuming we keep this models class (although I'd be against any magic like this as 90% of my development time is fighting against it :upside_down_face: )

File: src/adapters/interfaces/Logs.ts (hunk 3 above)
Comment (markusahlstrand) [reply src/adapters/interfaces/Logs.ts:10]: Yeah, this is a tough one. The omit types get a bit ugly in the docs. The other option is to do a base type..

File: src/models/User.ts (hunk 7 above)
Comment (markusahlstrand) [reply src/models/User.ts:427]: Yeah, this one is not obvious. We add the data parameter to the env object in the beginning of each request, but it's a different env compared to what's in the DO's..

**Review by markusahlstrand — COMMENTED** [review]
I think it looks good but think it might be better to do the logging outside the DO so we gradually could shut it down. It will solve the issue with the data not being available as well :)

File: src/routes/management-api/users.ts [hunk 8]
```diff
@@ -168,6 +166,13 @@ export class UsersMgmtController extends Controller {
       ...user,
       tenant_id: tenantId,
     });
+    const { tenant_id, id } = result;
+    await ctx.env.data.logs.create({
+      category: "update", // ooops! check all categories...
```
Comment (danielgent) [inline src/routes/management-api/users.ts:171]: bah, I was only copy-pasting the `message` :facepalm: 

I need to go back through these and check the `category`

**Comment by danielgent** [comment]
I can't get any logs for my user from auth0... I can get lots of the latest logs... but not per user

image


Ah well, I'll try again later

**Comment by danielgent** [comment]
Interestingly I can see them when requesting all the logs though...

image

File: package.json [hunk 9]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "build": "node tsoa.mjs",
     "dev": "node tsoa.mjs && wrangler dev --remote",
+    "dev:https": "node tsoa.mjs && wrangler dev --remote --local-protocol=https",
```
Comment (danielgent) [inline package.json:9]: happy not to add this but means I can point login2 locally to auth2, which does help with development :muscle:

File: src/migrate/migrations/2023-10-26T08:14:09_log_table.ts [hunk 10]
```diff
@@ -5,8 +5,11 @@ export async function up(db: Kysely<Database>): Promise<void> {
   await db.schema
     .createTable("logs")
     .addColumn("id", "varchar(255)", (col) => col.notNull().primaryKey())
-    .addColumn("client_id", "varchar(255)", (col) =>
-      col.references("clients.id").onDelete("cascade").notNull(),
+    .addColumn("tenant_id", "varchar(255)", (col) =>
+      col.references("tenants.id").onDelete("cascade").notNull(),
+    )
+    .addColumn("user_id", "varchar(255)", (col) =>
+      col.references("users.id").onDelete("cascade").notNull(),
```
Comment (danielgent) [inline src/migrate/migrations/2023-10-26T08:14:09_log_table.ts:12]: I could have done this as another migration but it's all applied now to dev :+1:

File: src/routes/management-api/logs.ts [hunk 11]
```diff
@@ -47,7 +46,6 @@ export class LogsController extends Controller {
     });
 
     if (include_totals) {
-      // don't like this type casting. copied from users route...
       return result as ListLogsResponse;
```
Comment (markusahlstrand) [inline src/routes/management-api/logs.ts:49]: Yeah, would be good to understand why it's needed..

**Review by markusahlstrand — APPROVED** [review]
Looks good

**Comment by danielgent** [comment]
> I think it looks good but think it might be better to do the logging outside the DO so we gradually could shut it down. It will solve the issue with the data not being available as well :)

that's what this PR does @markusahlstrand ! it moves the logging outside of the DO :wink: 

I'll merge this now

**Pull request merged**